### PR TITLE
adds an admission plugin to the sample apiserver.

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**:
adds an admission plugin to the sample apiserver.
the admission plugin checks whether `Flunder.Name` is not on the banned list.
including a unit test with various test scenarios.

**Special notes for your reviewer**:
https://github.com/kubernetes/kubernetes/issues/47868

**Release note**:

```
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+// The list is stored in Fischers API objects.
+func (d *disallowFlunder) Admit(a admission.Attributes) error {
+	// we are only interested in flunders
+	if a.GetKind().GroupKind() != api.Kind("Flunder") {
+		return nil
+	}
+
+	flunderName := a.GetName()
+	fischers, err := d.lister.List(labels.Everything())
+	if err != nil {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission.go]: Do we want to fail on each error ?

File: staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+var _ = wardleinitializer.WantsInternalWardleInformerFactory(&disallowFlunder{})
+
+// Admit ensures that the object in-flight is of kind Flunder.
+// In addition checks that the Name is not on the banned list.
+// The list is stored in Fischers API objects.
+func (d *disallowFlunder) Admit(a admission.Attributes) error {
+	// we are only interested in flunders
+	if a.GetKind().GroupKind() != api.Kind("Flunder") {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission.go]: alternatively we can check a `resource` not a `kind`

File: staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+	flunderName := a.GetName()
+	fischers, err := d.lister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, fischer := range fischers {
+		for _, disallowedFlunder := range fischer.DisallowedFlunders {
+			if flunderName == disallowedFlunder {
```
Comment (p0lyn0mial) [inline staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission.go]: we don't care about case sensitivity.

**Comment by p0lyn0mial** [comment]
I follow up wiring

File: staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
```
Comment (sttts) [inline staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission.go]: good that the fisher men know their fish by name.

File: staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission_test.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+			targetInitializer.Initialize(target)
+
+			err = admission.Validate(target)
+			if err != nil {
+				t.Fatalf("scenario %d: failed to initialize banflunder admission plugin due to =%v", index, err)
+			}
+
+			stop := make(chan struct{})
+			defer close(stop)
+			informersFactory.Start(wait.NeverStop)
```
Comment (sttts) [inline staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission_test.go]: not stop here as well?

**Comment by sttts** [comment]
One comment. Looks very good!

File: staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission_test.go (hunk 5 above)
Comment (p0lyn0mial) [inline staging/src/k8s.io/sample-apiserver/pkg/admission/plugin/banflunder/admission_test.go]: good point.

**Comment by p0lyn0mial** [comment]
@fejta Last time you did magic and all tests went green. I think I am still suffering from flakiness. Should I pin this PR somewhere ?

**Comment by p0lyn0mial** [comment]
@sttts updated and squashed, labels needed.

**Pull request merged**